community/ndp/runtime/src/main/java/org/neo4j/ndp/runtime/internal/ErrorTranslator.java

```diff
@@ -19,48 +19,94 @@
+import java.nio.charset.StandardCharsets;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.TimeZone;
+import java.util.UUID;
+
+import org.neo4j.helpers.Exceptions;
+import org.neo4j.kernel.Version;
 import org.neo4j.kernel.api.exceptions.Status;
+import org.neo4j.kernel.impl.logging.LogService;
 import org.neo4j.logging.Log;
 
+import static javax.xml.bind.DatatypeConverter.printBase64Binary;
+
 /**
  * Convert the mixed exceptions the underlying engine can throw to a cohesive set of known failures. This is an
  * intermediary mechanism.
  */
 public class ErrorTranslator
 {
+    private final Log internalLog;
+    private final Log userLog;
 
-    private final Log log;
+    public ErrorTranslator( LogService logging )
+    {
+        this(logging.getUserLog( ErrorTranslator.class ), logging.getInternalLog( ErrorTranslator.class ));
+    }
 
-    public ErrorTranslator( Log log )
+    public ErrorTranslator( Log userLog, Log internalLog )
     {
-        this.log = log;
+        this.internalLog = internalLog;
+        this.userLog = userLog;
     }
 
     public Neo4jError translate( Throwable any )
     {
-        if ( any instanceof KernelException )
-        {
-            return new Neo4jError( ((KernelException) any).status(), any.getMessage() );
-        }
-        else if ( any instanceof CypherException )
-        {
-            return new Neo4jError( ((CypherException) any).status(), any.getMessage() );
-        }
-        else
+        for( Throwable cause = any; cause != null; cause = cause.getCause() )
         {
-            if ( any.getCause() != null )
+            if ( cause instanceof Status.HasStatus )
             {
-                return translate( any.getCause() );
+                return new Neo4jError( ((Status.HasStatus) cause).status(), any.getMessage() );
             }
+        }
 
-            // Log unknown errors.
-            log.warn( "Client triggered unknown error: " + any.getMessage(), any );
+        // In this case, an error has "slipped out", and we don't have a good way to handle it. This indicates
+        // a buggy code path, and we need to try to convince whoever ends up here to tell us about it.
 
-            return new Neo4jError( Status.General.UnknownFailure, "An unexpected failure occurred: '"
-                                                                  + any.getMessage() + "'." );
-        }
+        // TODO: Perhaps move this whole block to somewhere more sensible?
+        String reference = UUID.randomUUID().toString();
+
+        // Log unknown errors.
+        userLog.error( String.format(
+                "Client triggered unexpected error. Help us fix this error by emailing the " +
+                "following report to issues@neotechnology.com: %n%s", generateReport( any, reference ) ), any );
+        internalLog.error( String.format( "Client triggered unexpected error, reference %s.", reference), any );
```

Comment:
If you're pulling both the userLog and internalLog from the internal kernel StoreLogService, then it's already setup to duplicate writes to the userLog into the internalLog (ref: https://github.com/neo4j/neo4j/blob/2.3/community/kernel/src/main/java/org/neo4j/kernel/impl/logging/SimpleLogService.java#L32 )